src/tactic/norm_cast.lean

```diff
@@ -25,225 +25,314 @@ try_for 1000 (mk_instance e)
 
-/--
-This is an attribute for simplification rules that are
-used to normalize casts.
+private meta def new_name (n : name) : name := name.mk_string "reversed" n
 
-Let r be = or ↔, then elimination lemmas of the shape
-Π ..., P ↑a1 ... ↑an r P a1 ... an should be given the
-attribute elim_cast.
+/--
+This is an attribute of the norm_cast tactic.
+It's intended for lemmas of the following shapes:
+  - Π ..., P ↑a1 ... ↑an = P a1 ... an
+  - Π ..., P ↑a1 ... ↑an ↔ P a1 ... an
 -/
 @[user_attribute]
 meta def elim_cast_attr : user_attribute simp_lemmas :=
 {
-    name      := `elim_cast,
-    descr     := "attribute for lemmas of the shape Π ..., P ↑a1 ... ↑an = P a1 ... an",
-    cache_cfg :=
-        { mk_cache     := mk_cache,
-          dependencies := [], },
+  name      := `elim_cast,
+  descr     := "attribute for lemmas of the shape Π ..., P ↑a1 ... ↑an = P a1 ... an",
+  cache_cfg :=
+    { mk_cache     := mk_cache,
+      dependencies := [], },
 }
 
-/--
-This is an attribute for simplification rules that are
-used to normalize casts.
+mk_simp_attribute push_cast "The `push_cast` simp attribute uses `squash_cast` and `move_cast` lemmas in the \"forward\" direction,
+to move casts toward the leaf nodes of the expression."
+
+private meta def aux_after_set (tac : expr → tactic (expr × (expr → expr))) :
+  expr → tactic (expr × (expr → expr))
+| (pi n bi d b) := do
+  uniq_n ← mk_fresh_name,
+  let b' := b.instantiate_var (local_const uniq_n n bi d),
+  (b', f) ← aux_after_set b',
+  return $ (
+    pi n bi d $ b'.abstract_local uniq_n,
+    λ e, lam n bi d $ ( f $ e (local_const uniq_n n bi d) ).abstract_local uniq_n
+  )
+| ty := tac ty
 
-Let r be = or ↔, then compositional lemmas of the shape
-Π ..., ↑(P a1 ... an) r P ↑a1 ... ↑an should be given the
-attribute move_cast.
+/-- Called after the `move_cast` attribute is applied to a declaration. -/
+private meta def move_cast_after_set (decl : name) (prio : ℕ) (pers : bool) : tactic unit :=
+do
+  (declaration.thm n l ty e) ← get_decl decl | failed,
+  let tac := λ ty, (flip_eq ty <|> flip_iff ty),
+  (ty', f) ← aux_after_set tac ty,
+  let e' := task.map f e,
+  let n' := new_name n,
+  add_decl (declaration.thm n' l ty' e'),
+  simp_attr.push_cast.set decl () tt
+
+/--
+This is an attribute of the norm_cast tactic.
+It's intended for lemmas of the following shapes:
+  - Π ..., ↑(P a1 ... an) = P ↑a1 ... ↑an
+  - Π ..., ↑(P a1 ... an) ↔ P ↑a1 ... ↑an
 -/
 @[user_attribute]
 meta def move_cast_attr : user_attribute simp_lemmas :=
 {
-    name      := `move_cast,
-    descr     := "attribute for lemmas of the shape Π ..., ↑(P a1 ... an) = P ↑a1 ... ↑an",
-    after_set := some after_set,
-    cache_cfg :=
-        { mk_cache     := mk_cache ∘ (list.map new_name),
-          dependencies := [], },
+  name      := `move_cast,
+  descr     := "attribute for lemmas of the shape Π ..., ↑(P a1 ... an) = P ↑a1 ... ↑an",
+  after_set := some move_cast_after_set,
```

Comment:
Could you add a `before_unset` argument? I noticed recently that we can't unset `move_cast` attributes because this is missing. I don't think it has to do anything, just `some tactic.skip` or some equivalent no-op.